File: Core/Components/Framework/NotificationBL.cs
using Bridge.Html5;
using Core.Clients;
using Core.Components.Extensions;
using Core.Components.Forms;
using Core.Enums;
using Core.Extensions;
using Core.Models;
using Core.MVVM;
using Core.ViewModels;
using System;
using System.Collections.Generic;
using System.Linq;
using ElementType = Core.MVVM.ElementType;

namespace Core.Components.Framework
{
    public class NotificationBL : EditableComponent
    {
        private const string NoMoreTask = "No more task";
        private static NotificationBL _instance;
        private static Observable<string> _countNtf;
        private static Observable<string> _countUser;
        private HTMLElement _task;
        private HTMLElement _countBadge;
        public static ObservableList<TaskNotification> Notifications { get; set; }
        public static ObservableList<User> UserActive { get; set; }
        private Token CurrentUser { get; set; }

        protected NotificationBL() : base(null)
        {
            Notifications = new ObservableList<TaskNotification>();
            UserActive = new ObservableList<User>();
            _countNtf = new Observable<string>();
            _countUser = new Observable<string>();
            Window.AddEventListener("task", ProcessIncomMessage);
        }

        public void ProcessIncomMessage(object obj)
        {
            if (obj is null)
            {
                return;
            }

            var task = (TaskNotification)obj;
            if (task is null)
            {
                return;
            }

            var existTask = Notifications.Data.FirstOrDefault(x => x.Id == task.Id);
            if (existTask == null)
            {
                Notifications.Add(task, 0);
                ToggleBageCount(Notifications.Data.Count);
            }
            SetBadgeNumber();
            var entity = Utils.GetEntityById(task.EntityId);
            task.Entity = new Entity { Name = entity.Name };
            /*@
            if (typeof(Notification) !== 'undefined' && Notification.permission === "granted") {
                this.ShowNativeNtf(task);
            } else if (typeof(Notification) !== 'undefined' && Notification.permission !== "denied") {
                Notification.requestPermission().then((permission) => {
                    if (permission !== 'granted') {
                        this.ShowToast(task);
                    }
                    else this.ShowNativeNtf(task);
                });
            } else this.ShowToast(task);
            */
        }

        private int SetBadgeNumber()
        {
            var unreadCount = Notifications.Data.Count(x => x.StatusId == ((int)TaskStateEnum.UnreadStatus).ToString());
            _countNtf.Data = unreadCount > 9 ? "9+" : unreadCount.ToString();
            var badge = unreadCount > 9 ? 9 : unreadCount;
            /*@
            if (typeof(cordova) !== 'undefined' &&
                typeof(cordova.plugins) !== 'undefined' &&
                typeof(cordova.plugins.notification) !== 'undefined') {
                cordova.plugins.notification.badge.requestPermission(function (granted) {
                    cordova.plugins.notification.badge.set(unreadCount);
                });
            }
            */
            return badge;
        }

        private void ShowNativeNtf(TaskNotification task)
        {
            if (task is null)
            {
                return;
            }

            /*@
            var nativeNtf = new Notification(task.Title,
            {
                body: task.Description,
                icon: task.Attachment,
                vibrate: [200, 100, 200],
                badge: "./favicon.ico"
            });
            nativeNtf.addEventListener('click', () => this.OpenNotification(task));
            setTimeout(() =>
            {
                nativeNtf.close();
            }, 7000);
            */
        }

        public static NotificationBL Instance
        {
            get
            {
                if (_instance is null)
                {
                    _instance = new NotificationBL();
                }
                return _instance;
            }
        }

        private static readonly HTMLElement NotiRoot = Document.GetElementById("notification-list");
        private static readonly HTMLElement ProfileRoot = Document.GetElementById("profile-info1");

        public override void Render()
        {
            Html.Take(NotiRoot).Clear();
            Html.Take("#user-active").Clear();
            SetBadgeNumber();
            CurrentUser = Client.Token;
            if (CurrentUser is null) return;
            CurrentUser.Avatar = (CurrentUser.Avatar.Contains("://") ? "" : Client.Origin) + (CurrentUser.Avatar.IsNullOrWhiteSpace() ? "./image/chinese.jfif" : CurrentUser.Avatar);
            RenderNotification();
            RenderProfile(ProfileRoot);
            var xhr = new XHRWrapper 
            {
                Url = $"/GetUserActive",
                Method = HttpMethod.POST,
                ShowError = false
            };
            Client.Instance.SubmitAsync<List<User>>(xhr)
                .Done(x => UserActive.Data = x)
                .Catch(Console.WriteLine);
        }

        public void RenderProfile(HTMLElement profileRoot)
        {
            if (profileRoot is null) return;
            var isSave = Window.LocalStorage.GetItem("isSave");
            Html.Take(profileRoot).Clear();
            var html = Html.Take(profileRoot);
            html.A.ClassName("navbar-nav-link d-flex align-items-center dropdown-toggle").DataAttr("toggle", "dropdown").Span.ClassName("text-truncate").Text(CurrentUser.TenantCode + ": " + CurrentUser.FullName).EndOf(ElementType.a)
                .Div.ClassName("dropdown-menu dropdown-menu-right notClose mt-0 border-0").Style("border-top-left-radius: 0;border-top-right-radius: 0")
                    .A.ClassName("dropdown-item").Event(EventType.Click, ViewProfile).I.ClassName("far fa-user").End.Text("Account (" + CurrentUser.TenantCode + ": " + CurrentUser.FullName + ")").EndOf(ElementType.a);
            html.Div.ClassName("dropdown-divider").EndOf(ElementType.div);
            var langSelect = new LangSelect(new Component(), html.GetContext());
            langSelect.Render();
            html.Div.ClassName("dropdown-divider").EndOf(ElementType.div);
            html.A.Event(EventType.Click, DarkOrLightModeSwitcher).ClassName("dropdown-item").I.ClassName("far fa-moon-cloud").End.Text("Dark/Light Mode").EndOf(ElementType.a);
            html.A.Event(EventType.Click, SignOut).ClassName("dropdown-item").I.ClassName("far fa-power-off").End.Text("Logout").EndOf(ElementType.a);
        }

        private void SignOut(Event e)
        {
            e.PreventDefault();
            var task = Client.Instance.PostAsync<bool>(Client.Token, "/user/signOut");
            task.Done(res =>
            {
                Client.SignOutEventHandler?.Invoke();
                Client.Token = null;
                EditForm.NotificationClient?.Close();
                Window.SetTimeout(() => Window.Location.Reload(), 1000);
            });
        }

        private void DarkOrLightModeSwitcher(Event e)
        {
            e.PreventDefault();
            var htmlElement = Document.DocumentElement;
            htmlElement.Style["filter"] = htmlElement.Style["filter"].ToString().Contains("(1)") ? "invert(0)" : "invert(1)";
        }

        private void ViewProfile(Event e)
        {
            e.PreventDefault();
            this.OpenTab(id: "User" + Client.Token.UserId,
                featureName: "UserProfile",
                factory: () =>
                {
                    var type = Type.GetType("Core.Fw.User.UserProfileBL");
                    var instance = Activator.CreateInstance(type) as TabEditor;
                    return instance;
                }).Done();
            Dispose();
        }

        private void RenderNotification()
        {
            if (NotiRoot is null) return;
            var html = Html.Take(NotiRoot).A.ClassName("navbar-nav-link").DataAttr("toggle", "dropdown").I.ClassName("far fa-bell fa-lg").EndOf(ElementType.i);
            if (_countNtf.Data != string.Empty)
            {
                html.Span.ClassName("badge badge-pill bg-warning-400 ml-auto ml-md-0").Text(_countNtf);
                _countBadge = Html.Context;
            };
            html.EndOf(ElementType.a);
            html.Div.Style("border-top-left-radius: 0;border-top-right-radius: 0")
                .ClassName("dropdown-menu dropdown-menu-right dropdown-content wmin-md-300 mt-0")
                .Style("border-top-left-radius: 0; border-top-right-radius: 0");
            html.Div.ForEach(Notifications, (task, index) =>
            {
                RenderTask(task);
            }).End.Render();
            html.A.ClassName("dropdown-item dropdown-footer").Event(EventType.Click, SeeMore).Text("See more").EndOf(ElementType.a);
        }

        private void RenderTask(TaskNotification task)
        {
            if (task is null)
            {
                return;
            }

            var className = task.StatusId == ((int)TaskStateEnum.UnreadStatus).ToString() ? "text-danger" : "text-muted";
            Html.Instance.A.ClassName("dropdown-item").Div.ClassName("media").Event(EventType.Click, (e) =>
            {
                OpenNotification(task);
            })
            .Div.ClassName("media-body").H3.ClassName("dropdown-item-title").Text(task.Title).Span.ClassName("float-right text-sm " + className).I.ClassName("fas fa-star").End.End.End
            .P.ClassName("text-sm").Text(task.Description).End
            .P.ClassName("text-sm text-muted")
                .I.ClassName("far fa-clock mr-1").End.Text(task.Deadline.ToString("dd/MM/yyyy HH:mm")).EndOf(ElementType.a);
        }

        private void ToggleBageCount(int count)
        {
            _countBadge.Style.Display = count == 0 ? Display.None : Display.InlineBlock;
        }

        private void SeeMore(Event e)
        {
            e.PreventDefault();
            var lastSeenTask = Notifications.Data.OrderByDescending(x => x.InsertedDate).FirstOrDefault();
            if (lastSeenTask is null) {
                Toast.Warning(NoMoreTask);
                return;
            }
            Spinner.AppendTo(Element, timeout: 250);
            var lastSeenDateStr = lastSeenTask.InsertedDate.ToISOFormat();
            var sql = new SqlViewModel
            {
                ComId = "Task",
                Action = "SeeMore",
                MetaConn = Client.MetaConn,
                DataConn = Client.DataConn,
                Params = JSON.Stringify(new { Date = lastSeenDateStr })
            };
            Client.Instance.UserSvc(sql).Done(ds =>
            {
                var olderItems = ds.Length > 0 ? ds[0].Select(x => x.CastProp<TaskNotification>()).ToArray() : null;
                if (olderItems.Nothing()) 
                {
                    Toast.Warning(NoMoreTask);
                    return;
                }
                var taskList = Notifications.Data.Union(olderItems).ToList();
                Notifications.Data = taskList;
            }).Catch(err => {
                Toast.Warning(err?.Message ?? NoMoreTask);
            });
        }

        public void MarkAllAsRead(Event e)
        {
            e.PreventDefault();
            Client.Instance.PostAsync<bool>(null, nameof(TaskNotification) + "/MarkAllAsRead")
            .Done(res =>
            {
                ToggleBageCount(Notifications.Data.Count(x => x.StatusId == ((int)TaskStateEnum.UnreadStatus).ToString()));
                foreach (var task in _task.QuerySelectorAll(".task-unread"))
                {
                    task.ReplaceClass("task-unread", "task-read");
                }
            });
        }

        public void OpenNotification(TaskNotification notification)
        {
            MarkAsRead(notification);
        }

        protected override void RemoveDOM()
        {
            Html.Take("#notification").Clear();
        }

        private void MarkAsRead(TaskNotification task)
        {
            task.StatusId = ((int)TaskStateEnum.Read).ToString();
            Client.Instance.PatchAsync(task.MapToPatch())
                .Done(x => SetBadgeNumber());
        }

        public override void Dispose()
        {
            _task.AddClass("hide");
        }
    }
}

File: CoreAPI/wwwRoot/js/core/components/framework/notificationBL.js
import { Client } from "clients/client";
import EditableComponent from "./editableComponent.js";
import { WebSocketClient } from "clients/websocketClient";
import { KeyCodeEnum, RoleEnum } from "models/enum";
import EventType from "models/eventType";
import { PopupEditor } from "popupEditor";
import { Toast } from "toast";
import { Html } from "utils/html";
import { Utils } from "utils/utils";
import { TaskStateEnum } from "Core/Enums"


export class NotificationBL extends EditableComponent {
    static NoMoreTask = "No more task";
    static _instance = null;
    static _countNtf = null;
    static _countUser = null;
    static Notifications = [];
    static UserActive = [];
    static NotiRoot = document.getElementById("notification-list");
    static ProfileRoot = document.getElementById("profile-info1");

    constructor() {
        super(null);
        this.notifications = [];
        this.userActive = [];
        this._countNtf = "";
        this._countUser = "";
        this._task = null;
        this._countBadge = null;
        this.currentUser = null;
        window.addEventListener("task", this.processIncomMessage.bind(this));
    }

    processIncomMessage(event) {
        const obj = event.detail;
        if (!obj) {
            return;
        }

        const task = obj;
        if (!task) {
            return;
        }

        const existTask = this.notifications.find(x => x.Id === task.Id);
        if (!existTask) {
            this.notifications.push(task);
            this.toggleBadgeCount(this.notifications.length);
        }
        this.setBadgeNumber();
        const entity = Utils.getEntityById(task.EntityId);
        task.Entity = { Name: entity.Name };

    //     if (typeof(Notification) !== 'undefined' && Notification.permission === "granted") {
    //         this.showNativeNtf(task);
    //     } else if (typeof(Notification) !== 'undefined' && Notification.permission !== "denied") {
    //         Notification.requestPermission().then((permission) => {
    //             if (permission !== 'granted') {
    //                 this.showToast(task);
    //             } else {
    //                 this.showNativeNtf(task);
    //             }
    //         });
    //     } else {
    //         this.showToast(task);
    //     }
    // }

    setBadgeNumber() {
        const unreadCount = this.notifications.filter(x => x.StatusId === TaskStateEnum.UnreadStatus.toString()).length;
        this._countNtf = unreadCount > 9 ? "9+" : unreadCount.toString();
        const badge = unreadCount > 9 ? 9 : unreadCount;

        // if (typeof(cordova) !== 'undefined' &&
        //     typeof(cordova.plugins) !== 'undefined' &&
        //     typeof(cordova.plugins.notification) !== 'undefined') {
        //     cordova.plugins.notification.badge.requestPermission(function(granted) {
        //         cordova.plugins.notification.badge.set(unreadCount);
        //     });
        // }
        return badge;
    }

    // showNativeNtf(task) {
    //     if (!task) {
    //         return;
    //     }

    //     const nativeNtf = new Notification(task.Title, {
    //         body: task.Description,
    //         icon: task.Attachment,
    //         vibrate: [200, 100, 200],
    //         badge: "./favicon.ico"
    //     });

    //     nativeNtf.addEventListener('click', () => this.openNotification(task));
    //     setTimeout(() => {
    //         nativeNtf.close();
    //     }, 7000);
    // }

    // showToast(task) {
    //     // Implementation for showing toast notification
    // }

    // toggleBadgeCount(count) {
    //     // Implementation for toggling badge count
    // }

    // openNotification(task) {
    //     // Implementation for opening notification
    // }

    static getInstance() {
        if (!this._instance) {
            this._instance = new NotificationBL();
        }
        return this._instance;
    }

    render() {
        const notiRoot = document.getElementById("notification-list");
        const profileRoot = document.getElementById("profile-info1");

        if (!notiRoot || !profileRoot) {
            return;
        }

        notiRoot.innerHTML = '';
        document.querySelector("#user-active").innerHTML = '';
        this.setBadgeNumber();

        this.currentUser = Client.token;
        if (!this.currentUser) return;

        this.currentUser.Avatar = (this.currentUser.Avatar.includes("://") ? "" : Client.origin) + (this.currentUser.Avatar.trim() === "" ? "./image/chinese.jfif" : this.currentUser.Avatar);
        this.renderNotification();
        this.renderProfile(profileRoot);

        const xhr = {
            url: "/GetUserActive",
            method: "POST",
            showError: false
        };

        Client.instance.submitAsync(xhr)
            .then(users => this.userActive = users)
            .catch(console.log);
    }

    renderProfile(profileRoot) {
        if (!profileRoot) return;

        const isSave = window.localStorage.getItem("isSave");
        profileRoot.innerHTML = '';

        const html = new HtmlElementWrapper(profileRoot);
        html.addClassName("navbar-nav-link d-flex align-items-center dropdown-toggle")
            .setDataAttr("toggle", "dropdown")
            .appendSpan("text-truncate", this.currentUser.TenantCode + ": " + this.currentUser.FullName)
            .end(ElementType.a)
            .appendDiv("dropdown-menu dropdown-menu-right notClose mt-0 border-0", "border-top-left-radius: 0; border-top-right-radius: 0")
            .appendA("dropdown-item", this.viewProfile.bind(this))
            .appendI("far fa-user")
            .end("Account (" + this.currentUser.TenantCode + ": " + this.currentUser.FullName + ")")
            .end(ElementType.a)
            .appendDiv("dropdown-divider")
            .end(ElementType.div);

        const langSelect = new LangSelect(new Component(), html.getContext());
        langSelect.render();

        html.appendDiv("dropdown-divider")
            .end(ElementType.div)
            .appendA("dropdown-item", this.darkOrLightModeSwitcher.bind(this))
            .appendI("far fa-moon-cloud")
            .end("Dark/Light Mode")
            .end(ElementType.a)
            .appendA("dropdown-item", this.signOut.bind(this))
            .appendI("far fa-power-off")
            .end("Logout")
            .end(ElementType.a);
    }

    signOut(event) {
        event.preventDefault();
        const task = Client.instance.postAsync(Client.token, "/user/signOut");
        task.then(res => {
            Client.signOutEventHandler?.();
            Client.token = null;
            EditForm.notificationClient?.close();
            setTimeout(() => {
                window.location.reload();
            }, 1000);
        });
    }

    darkOrLightModeSwitcher(event) {
        event.preventDefault();
        const htmlElement = document.documentElement;
        htmlElement.style.filter = htmlElement.style.filter.includes("(1)") ? "invert(0)" : "invert(1)";
    }

    viewProfile(event) {
        event.preventDefault();
        this.openTab({
            id: "User" + Client.token.UserId,
            featureName: "UserProfile",
            factory: () => {
                const type = Type.getType("Core.Fw.User.UserProfileBL");
                const instance = new type();
                return instance;
            }
        }).then(() => {
            this.dispose();
        });
    }

    renderNotification() {
        const notiRoot = NotificationBL.NotiRoot;
        if (!notiRoot) return;

        const html = new HtmlElementWrapper(notiRoot);
        html.addClassName("navbar-nav-link")
            .setDataAttr("toggle", "dropdown")
            .appendI("far fa-bell fa-lg")
            .end(ElementType.i);

        if (this._countNtf !== '') {
            html.appendSpan("badge badge-pill bg-warning-400 ml-auto ml-md-0", this._countNtf);
            this._countBadge = html.getContext();
        }

        html.end(ElementType.a)
            .appendDiv("dropdown-menu dropdown-menu-right dropdown-content wmin-md-300 mt-0", "border-top-left-radius: 0; border-top-right-radius: 0")
            .forEach(this.notifications, (task, index) => {
                this.renderTask(task);
            })
            .end()
            .appendA("dropdown-item dropdown-footer", this.seeMore.bind(this), "See more")
            .end(ElementType.a);
    }

    renderTask(task) {
        if (!task) {
            return;
        }

        const className = task.StatusId === TaskStateEnum.UnreadStatus.toString() ? "text-danger" : "text-muted";
        const html = new HtmlElementWrapper();

        html.appendA("dropdown-item")
            .appendDiv("media", (event) => {
                this.openNotification(task);
            })
            .appendDiv("media-body")
            .appendH3("dropdown-item-title", task.Title)
            .appendSpan("float-right text-sm " + className)
            .appendI("fas fa-star")
            .end()
            .end()
            .appendP("text-sm", task.Description)
            .end()
            .appendP("text-sm text-muted")
            .appendI("far fa-clock mr-1")
            .end(task.Deadline.toLocaleDateString("en-GB") + " " + task.Deadline.toLocaleTimeString("en-GB"))
            .end(ElementType.a);
    }

    toggleBadgeCount(count) {
        this._countBadge.style.display = count === 0 ? 'none' : 'inline-block';
    }

    seeMore(event) {
        event.preventDefault();
        const lastSeenTask = this.notifications.sort((a, b) => new Date(b.InsertedDate) - new Date(a.InsertedDate))[0];
        if (!lastSeenTask) {
            Toast.warning(NotificationBL.NoMoreTask);
            return;
        }
        Spinner.appendTo(Element, 250);
        const lastSeenDateStr = new Date(lastSeenTask.InsertedDate).toISOString();
        const sql = {
            ComId: "Task",
            Action: "SeeMore",
            MetaConn: Client.MetaConn,
            DataConn: Client.DataConn,
            Params: JSON.stringify({ Date: lastSeenDateStr })
        };
        Client.instance.userSvc(sql).then(ds => {
            const olderItems = ds.length > 0 ? ds[0].map(x => new TaskNotification(x)) : null;
            if (!olderItems || olderItems.length === 0) {
                Toast.warning(NotificationBL.NoMoreTask);
                return;
            }
            const taskList = [...this.notifications, ...olderItems];
            this.notifications = taskList;
        }).catch(err => {
            Toast.warning(err?.message ?? NotificationBL.NoMoreTask);
        });
    }

    markAllAsRead(event) {
        event.preventDefault();
        Client.instance.postAsync(null, `${TaskNotification.name}/MarkAllAsRead`).then(res => {
            this.toggleBadgeCount(this.notifications.filter(x => x.StatusId === TaskStateEnum.UnreadStatus.toString()).length);
            document.querySelectorAll(".task-unread").forEach(task => {
                task.classList.replace("task-unread", "task-read");
            });
        });
    }

    openNotification(notification) {
        this.markAsRead(notification);
    }

    removeDOM() {
        document.querySelector("#notification").innerHTML = '';
    }

    markAsRead(task) {
        task.StatusId = TaskStateEnum.Read.toString();
        Client.instance.patchAsync(task.toPatch()).then(x => {
            this.setBadgeNumber();
        });
    }

    dispose() {
        this._task.classList.add("hide");
    }
}
